samza-sql/src/main/java/org/apache/samza/sql/planner/SamzaSqlValidator.java

```diff
@@ -0,0 +1,279 @@
+import org.apache.calcite.rel.logical.LogicalProject;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactoryImpl;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rel.type.RelRecordType;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.Config;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+import org.apache.samza.sql.dsl.SamzaSqlDslConverter;
+import org.apache.samza.sql.interfaces.RelSchemaProvider;
+import org.apache.samza.sql.interfaces.SamzaSqlJavaTypeFactoryImpl;
+import org.apache.samza.sql.runner.SamzaSqlApplicationConfig;
+import org.apache.samza.sql.util.SamzaSqlQueryParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * SamzaSqlValidator that uses calcite engine to convert the sql query to relational graph and validates the query
+ * including the output.
+ */
+public class SamzaSqlValidator {
+  private static final Logger LOG = LoggerFactory.getLogger(SamzaSqlValidator.class);
+
+  private final Config config;
+
+  public SamzaSqlValidator(Config config) {
+    this.config = config;
+  }
+
+  /**
+   * Validate a list of sql statements
+   * @param sqlStmts list of sql statements
+   * @throws SamzaSqlValidatorException
+   */
+  public void validate(List<String> sqlStmts) throws SamzaSqlValidatorException {
+    SamzaSqlApplicationConfig sqlConfig = SamzaSqlDslConverter.getSqlConfig(sqlStmts, config);
+    QueryPlanner planner = SamzaSqlDslConverter.getQueryPlanner(sqlConfig);
+
+    for (String sql: sqlStmts) {
+      // we always pass only select query to the planner for samza sql. The reason is that samza sql supports
+      // schema evolution where source and destination could up to an extent have independent schema evolution while
+      // calcite expects strict conformance of the destination schema with that of the fields in the select query.
+      SamzaSqlQueryParser.QueryInfo qinfo = SamzaSqlQueryParser.parseQuery(sql);
+      RelRoot relRoot;
+      try {
+        relRoot = planner.plan(qinfo.getSelectQuery());
+      } catch (SamzaException e) {
+        throw new SamzaSqlValidatorException("Calcite planning for sql failed.", e);
+      }
+
+      // Now that we have logical plan, validate different aspects.
+      validate(relRoot, qinfo, sqlConfig);
+    }
+  }
+
+  protected void validate(RelRoot relRoot, SamzaSqlQueryParser.QueryInfo qinfo, SamzaSqlApplicationConfig sqlConfig)
+      throws SamzaSqlValidatorException {
+    // Validate select fields (including Udf return types) with output schema
+    validateOutput(relRoot, sqlConfig.getRelSchemaProviders().get(qinfo.getSink()));
+
+    // TODO:
+    //  1. Validate Udf arguments.
+    //  2. Validate operators. These are the operators that are supported by Calcite but not by Samza Sql.
+    //     Eg: LogicalAggregate is not supported by Samza Sql.
```

Comment:
I am not sure i understand this. Are you saying we should add support for validating aggregation  when samza SQL supports aggregations?

Reply:
Changed to "LogicalAggregate with sum function is not supported by Samza Sql."